### PR TITLE
Make NUnitLite use AssertionResult info in it's reporting

### LINKED ISSUE #1889: Modify nunitlite to display multiple assert information
Currently, the nunitlite runner uses the single failure message created for compatibility in order to display the result of a test using multiple asserts. It should use the actual result of each assert to display both the message and stack frame.

### PR DESCRIPTION
**Pull request by CharliePoole**
Fixes #1889 

The first commit is an initial pass at using AssertionResult for our reporting. Before merging, let's use this as an opportunity to do a final review on the formatting of our reports.

In order to make this change I did some refactoring of StackFilter so it could be reused for different kinds of filtering.

### CODE REVIEW

**Comment by CharliePoole** [comment]
Here's an example of output when we were using the "compatible" `<failure>` element:

```
Errors and Failures

1) Failed : NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo
  Multiple Assert block had 2 failure(s).

  1)   RealPart
  Expected: 5.0d
  But was:  5.2000000000000002d


  2)   ImaginaryPart
  Expected: 4.2000000000000002d
  But was:  3.8999999999999999d



at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo()
```

Here's the same test using the new `<assertion>` elements:

```
Errors and Failures

1) Failed : NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo
  RealPart
  Expected: 5.0d
  But was:  5.2000000000000002d
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0()
at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in D:\Dev\NUnit\nunit-3.0\src\NUnitFramework\framework\Assert.cs:line 290
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo()

2) Failed : NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo
  ImaginaryPart
  Expected: 4.2000000000000002d
  But was:  3.8999999999999999d
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0()
at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in D:\Dev\NUnit\nunit-3.0\src\NUnitFramework\framework\Assert.cs:line 290
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo()
```

Essentially, the new way, you get two entries for the same test. The count of failures is based on tests, not asserts, so the summary would indicate 1 failure in both cases. Does anyone think that's a problem?

Let me know if you think the layout should be in any way different - right now it is exactly the same as for tests that do not use `Assert.Multiple`.

**Comment by rprouse** [comment]
>  The count of failures is based on tests, not asserts, so the summary would indicate 1 failure in both cases. Does anyone think that's a problem?

I am okay with that. One idea to make it clearer might be for multiple results to use a two part number. It would require refactoring the display results. It could look like this,

```
Errors and Failures

1-1) Failed : NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo
  RealPart
  Expected: 5.0d
  But was:  5.2000000000000002d
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0()
at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in D:\Dev\NUnit\nunit-3.0\src\NUnitFramework\framework\Assert.cs:line 290
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo()

1-2) Failed : NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo
  ImaginaryPart
  Expected: 4.2000000000000002d
  But was:  3.8999999999999999d
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.<>c.<MultipleAssertFailureDemo>b__3_0()
at NUnit.Framework.Assert.Multiple(TestDelegate testDelegate) in D:\Dev\NUnit\nunit-3.0\src\NUnitFramework\framework\Assert.cs:line 290
at NUnit.Framework.Assertions.Tests.AssertMultipleTests.MultipleAssertFailureDemo()
```

It would also be nice to clean up the extra newlines in the the "compatible" `<failure>` element so that older runners look okay. It looks like we have two or three newlines between blocks.

Other than that, looks good. Feel free to merge when ready.

**Comment by CharliePoole** [comment]
@rprouse I like your idea. I thought about a further refinement which would eliminate the replication of the first line, but I think that's for the future, when we may have mixed assertion results listed under a test and some of the statuses don't match the overall status of the test.

File: src/NUnitFramework/nunitlite/TextUI.cs [hunk 1]
```diff
@@ -482,19 +482,40 @@ private void DisplayNotRunResults(ITestResult result)
...
         private void DisplayTestResult(ITestResult result)
         {
-            string status = result.ResultState.Label;
+            ResultState resultState = result.ResultState;
+            string fullName = result.FullName;
+            string message = result.Message;
+            string stackTrace = result.StackTrace;
+            string reportID = (++_reportIndex).ToString();
+
+            //if (result.AssertionResults.Count > 0)
```
Comment (rprouse) [inline src/NUnitFramework/nunitlite/TextUI.cs]: I assume you didn't intend to keep this commented out. Is there another push coming?

Comment (CharliePoole) [inline src/NUnitFramework/nunitlite/TextUI.cs]: You're right and there is! I'm just experimenting with the "compatible" display there.

**Pull request merged**